sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala

```diff
@@ -85,17 +85,18 @@ sealed class Metadata private[types] (private[types] val map: Map[String, Any])
   override def equals(obj: Any): Boolean = {
     obj match {
       case that: Metadata =>
-        if (map.keySet == that.map.keySet) {
```

Comment:
This old implementation of `Metadata.equals()` was insanely inefficient because `map.keySet()` performs tons of object allocation. The old code also used `array.view` to compare array contents for equality, which is going to be slower than `Arrays.equals()`.